### ISSUE TITLE
Bug in extinction curve waveset unit conversion?

### ISSUE THREAD

**Issue by pllim** [spacetelescope/pysynphot#8]
In the following case, I have an extinction curve and a source spectrum. Extinction curve has wavelength unit of inverse micron by default, instead of the regular Angstrom unit:

```python
>>> from pysynphot.spparser import parse_spec
>>> red1 = parse_spec('ebmvx(0.04,gal3)')
>>> red1.waveunits
<pysynphot.units.InverseMicron at ...>
>>> red1.GetWaveSet()
array([  0.44999999,   0.54549998,   0.64099997, ...,
         9.61800003,   9.71350002,   9.80900002,   9.90450001,  10.        ], dtype=float32)
>>> red1._wavetable
array([ 22222.22265625,  18331.80664062,  15600.625     , ...,
         1039.71716309,   1029.49499512,   1019.47192383,   1009.64208984,
         1000.        ], dtype=float32)
>>> sp1 = parse_spec('rn(spec($PYSYN_CDBS/grid/pickles/dat_uvk/pickles_uk_100.fits),band(cousins,i),28.0,vegamag)')
>>> sp1.waveunits
<pysynphot.units.Angstrom at ...>
>>> sp1.GetWaveSet()
array([  1150.,   1155.,   1160., ...,  24990.,  24995.,  25000.], dtype=float32)
```

When I multiply the extinction curve to source spectrum, theoretically, both their wavesets would have been combined in Angstrom internally. So, from the outputs above, I am expecting a combined waveset that runs from 1000 to 25000 Angstrom. But instead, as shown below, the composite spectrum failed to convert inverse micron to Angstrom prior to merging:

```python
>>> sp2 = sp1 * red1
>>> sp2.waveunits
<pysynphot.units.Angstrom at ...>
>>> sp2.wave
array([  4.49999988e-01,   5.45499980e-01,   6.40999973e-01, ...,
         2.49900000e+04,   2.49950000e+04,   2.50000000e+04], dtype=float32)
```

This looks like a bug to me. @vglaidler , any ideas?

**Comment by pllim** [comment]
I traced this to inconsistent treatment of `GetWaveSet()` methods across different classes. In the following code snippet, `SpectralElement` returns waveset in user unit (which is what the extinction curve is doing), while `TabularSourceSpectrum` returns waveset in internal unit (which is what the source spectrum is doing). However, when the two are multiplied, `CompositeSourceSpectrum` does not take that into account. There are many ways to "fix" this but I am not sure what was the original intended behavior here:

```python
class SpectralElement(Integrator):
    def GetWaveSet(self):
        """Obtain the wavelength set for the spectrum.

        Returns
        -------
        wave : array_like
            Wavelength set in user unit.

        """
        wave = units.Angstrom().Convert(self._wavetable, self.waveunits.name)
        return wave


class TabularSourceSpectrum(SourceSpectrum):
    def GetWaveSet(self):
        """Return the wavelength set for the spectrum.

        Returns
        -------
        waveset : array_like
            Wavelength set (a copy of the internal wavelength table).

        """
        # For a TabularSource Spectrum, the WaveSet is just the _wavetable
        # member.  Return a copy so that there is no reference to the original
        # object.
        return self._wavetable.copy()


class CompositeSourceSpectrum(SourceSpectrum):
    def GetWaveSet(self):
        """Obtain the wavelength set for the composite spectrum.
        This is done by using :func:`MergeWaveSets` to form a union of
        wavelength sets from its components.

        Returns
        -------
        waveset : array_like
            Composite wavelength set.

        """
        waveset1 = self.component1.GetWaveSet()
        waveset2 = self.component2.GetWaveSet()
        return MergeWaveSets(waveset1, waveset2)
```

**Comment by vglaidler** [comment]
Do you get a different result if you put the extinction into the parser, rather than as a separate step? That's how we use it in the HST ETC, eg

    'source_expr'  : 'rn(unit(1.,flam) * ebmvx(3.000000,mwavg),band(johnson,v),15,vegamag)',

**Comment by vglaidler** [comment]
I would also dig into the history of when those two GetWaveSet routines were written & whether there's anything in the commit messages that would be helpful in understanding intent.

**Comment by pllim** [comment]
> Do you get a different result if you put the extinction into the parser, rather than as a separate step?

Nope. I see the same bug.

```python
>>> from pysynphot.spparser import parse_spec
>>> sp = parse_spec('rn(unit(1.,flam) * ebmvx(3.000000,mwavg),band(johnson,v),15,vegamag)')
>>> sp.wave
array([  4.49999988e-01,   5.45499980e-01,   6.40999973e-01, ...,
         2.59691986e+04,   2.59794616e+04,   2.59897288e+04])
```

**Comment by pllim** [comment]
>  dig into the history of when those two GetWaveSet routines were written

For `TabularSourceSpectrum` and `CompositeSourceSpectrum`, it was traced back to 9c0f3a9 by @vglaidler on Jul 21, 2008. Here is the full commit message:

```
 Major refactoring to support renormalization (#70):

  - the spectrum.renormalize function has been removed
  - the spectrum.Spectrum.renorm() method calls the new function
  renorm.StdRenorm() 
  - the spparser code has been modified to call the method on the
  spectrum
  - Pass/fail criteria have been loosened slightly for some tests in
  the pysynphot test suite; the loosest criterion is now 2.5%. 
  - A small set of tests that performs a direct comparison between
  synphot and pysynphot was added (renorm_test.py). This set has not
  yet been included into the tests that are run nightly, however.
  - Detritus from previous attempts at renormalization have been
  removed, notably the Spectrum.photonrate() and the FluxUnits.perphoton

The obsolete magnitudes.py file was also removed (#29).

For more detail, see the revision log for 
http://astropy.scipy.org/svn/astrolib/branches/pysynphot_renorm.


git-svn-id: https://aeon.stsci.edu/ssb/svn/astrolib/trunk/pysynphot@536 90a0a646-be8a-0410-bb88-9290da87bc01
```

For `SpectralElement`, it was traced back to b419586 by @vglaidler on Jul 25, 2008. Here is the full commit message:

```
 #104: support unit conversion of SpectralElements to non-Angstrom waveunits

git-svn-id: https://aeon.stsci.edu/ssb/svn/astrolib/trunk/pysynphot@548 90a0a646-be8a-0410-bb88-9290da87bc01
```

Vicki, do any of these jog your memory?

**Comment by vglaidler** [comment]
Doesn't ring a bell, but I poked around in the timeline some and I see that this work was done before some of pysynphot's basic infrastructure, and certainly before we got very careful about waveset merging - Matt did much of that years later. We mostly did that in the context of the Observation class though. 
  -- oh there's a followup question: what happens when you make an Observation of the resulting CompositeSourceSpectrum? Does the problem get corrected?

That leads me to suggest that you should check all getWaveSet methods on all pysynphot objects and check for consistency. If this is the one outlier, then change it to match the others. *However*, remember that this is a base class, so that might have unanticipated ripple effects that should be broadly tested.

Here's some additional history: [references to extinction in tickets&changesets](https://aeon.stsci.edu/ssb/trac/astrolib/search?changeset=on&milestone=on&ticket=on&wiki=on&q=extinction&&noquickjump=1)

One last question: can you reproduce this with the astrolib version of pysynphot?

**Comment by pllim** [comment]
> can you reproduce this with the astrolib version of pysynphot?

Not sure if I understand this question. This *is* the `astrolib` version of PySynphot.

**Comment by vglaidler** [comment]
>>    can you reproduce this with the astrolib version of pysynphot?

>Not sure if I understand this question. This is the astrolib version of PySynphot.

Ah, I wasn't tracking that carefully & don't know what if any changes had been made here.

**Comment by pllim** [comment]
> you should check all getWaveSet methods 

This is a great suggestion. I think I found a similar bug also in `GaussianSource`. See below:

```python
>>> g = S.GaussianSource(1e-23, 500, 10, waveunits='nm', fluxunits='jy')
>>> g.waveunits
<pysynphot.units.Nm at 0x7f2f1d45d4a8>
>>> g.GetWaveSet()  # docstring suggests this is in user unit, and it is
array([ 478.76695499,  479.19161589,  479.61627679, ...,
        519.53440141,  519.95906231,  520.38372321,  520.80838411,
        521.23304501])
```

The problem here is that in its parent class, `.wave` set by `getArrays()` assumes that `GetWaveSet()` is always in Angstrom. As a result, the following does not make sense:

```python
>>> g.wave  # What unit is this???
array([ 47.8766955 ,  47.91916159,  47.96162768, ...,
        51.95344014,  51.99590623,  52.03837232,  52.08083841,  52.1233045 ])
```

The rest of the `GetWaveSet()` implementations (aside from the two highlighted classes) have implicit assumptions of always being in Angstrom unit.

**Comment by pllim** [comment]
> what happens when you make an Observation of the resulting CompositeSourceSpectrum? Does the problem get corrected?

The answer is no, as illustrated below. With the exception that if you only use binned data from a bandpass with a pre-defined "wavetable", then you are not affected.

```python
>>> sp = parse_spec('rn(unit(1.,flam) * ebmvx(3.000000,mwavg),band(johnson,v),15,vegamag)')
>>> bp = S.ObsBandpass('acs,wfc1,f814w')  # has a wavetable
>>> obs = S.Observation(sp, bp)
>>> obs.wave  # natural (bad)
array([  4.49999988e-01,   5.45499980e-01,   6.40999973e-01, ...,
         2.59897288e+04,   3.00000000e+04,   3.00100000e+04])
>>> obs.binwave   # binned (good)
array([  1000.,   1001.,   1002., ...,  10998.,  10999.,  11000.])
>>> bp2 = S.Box(5000, 1000)  # no wavetable
>>> obs2 = S.Observation(sp, bp2)
>>> obs2.wave  # natural (bad)
array([  4.49999988e-01,   5.45499980e-01,   6.40999973e-01, ...,
         2.59691986e+04,   2.59794616e+04,   2.59897288e+04])
>>> obs2.binwave  # binned (bad)
array([  4.49999988e-01,   5.45499980e-01,   6.40999973e-01, ...,
         2.59691986e+04,   2.59794616e+04,   2.59897288e+04])
```

**Comment by pllim** [comment]
> don't know what if any changes had been made here

I haven't made any big changes to this code base since I took over from Matt. My work mostly is in refactoring it for Astropy, which happens in another repo.

**Comment by pllim** [comment]
Note: A symptom of this bug might have also manifested in https://aeon.stsci.edu/ssb/trac/astrolib/ticket/169

**Comment by pllim** [comment]
Fixed and released as 0.9.8.4 in AstroConda.
